activerecord/lib/active_record/connection_adapters/schema_cache.rb

```diff
@@ -129,7 +129,16 @@ def marshal_dump
 
       def marshal_load(array)
         @version, @columns, @columns_hash, @primary_keys, @data_sources, @indexes, @database_version = array
-        @indexes = @indexes || {}
+        @indexes ||= {}
+
+        @columns = @columns.map { |table_name, columns| [-table_name, columns.map(&:deduplicate)] }.to_h
+        @columns_hash = @columns_hash.map do |table_name, columns|
+          columns_hash = columns.map { |column_name, column| [-column_name, column.deduplicate] }.to_h
+          [-table_name, columns_hash]
+        end.to_h
+        @primary_keys = @primary_keys.map { |table_name, column_name| [-table_name, -column_name] }.to_h
+        @data_sources = @data_sources.transform_keys { |table_name| -table_name }
+        @indexes = @indexes.map { |table_name, indexes| [-table_name, indexes] }.to_h
```

Comment:
I think we should make `IndexDefinition` Deduplicable as well, but it doesn't have a `==` nor `hash` method, so I think it would be better handled in a followup.

Reply:
Agreed on `IndexDefinition`!

I find this code quite hard to parse. The meaning is hidden behind the processing and the individual `-` are lost — plus the differences for when to use `deduplicate`. Would something like this be clearer?

```ruby
  @columns      = deep_deduplicate(@columns) { |columns| columns.map(&:deduplicate) }
  @columns_hash = deep_deduplicate(@columns_hash) { |columns| deep_deduplicate(columns, &:deduplicate) }
  @primary_keys = deep_deduplicate(@primary_keys)
  @data_sources = deep_deduplicate(@data_sources, &:itself)
  @indexes      = deep_deduplicate(@indexes, &:itself)
end

def deep_deduplicate(values, &block)
  block ||= :-@.to_proc
  values.transform_keys(&:-@).transform_values(&block)
end
```

The `:-@.to_proc` should probably be cleaned up a bit somehow. Also not sure how wasteful it is to loop through the probably gargantuan hashes twice.

Another thing we could do is make the deduplicateable objects respond to `-@` instead of `deduplicate`. Then we don't have to make the difference between them and follow Ruby's new semantics.

Reply:
> Another thing we could do is make the deduplicateable objects respond to `-@` instead of `deduplicate`. Then we don't have to make the difference between them and follow Ruby's new semantics.

That's what I wanted to do initially, however I though that since numeric types respond to that same interface for a very different meaning made me reconsider it. So I'm split on this.

Reply:
> Agreed on IndexDefinition!

Actually I wonder about these. Turns out our internal SchemaCache implementation doesn't persist them, our prod workers have 0 of them in memory. So I wonder how useful they actually are.

Reply:
iirc, indexes were added to the cache in Rails 6.